pkg/steps/template.go

```diff
@@ -421,18 +421,50 @@ func createOrRestartPod(podClient coreclientset.PodInterface, pod *coreapi.Pod)
 	return created, nil
 }
 
-func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
+func waitForTeardownToComplete(pod *coreapi.Pod) {
 	timeout := 300
 	for i := 0; i < timeout; i += 2 {
-		pod, err := podClient.Get(name, meta.GetOptions{})
-		if errors.IsNotFound(err) {
-			return nil
-		}
-		if err != nil {
-			return fmt.Errorf("could not retrieve deleting pod: %v", err)
+		if !podTeardownCompleted(pod) {
+			log.Printf("Waiting for pod %s to complete teardown ... (%ds/%d)", pod.GetName(), i, timeout)
+			time.Sleep(2 * time.Second)
+		} else {
+			break
 		}
-		if pod.UID != uid {
-			return nil
+	}
+}
+
+func getPodByName(podClient coreclientset.PodInterface, name string, uid types.UID) (*coreapi.Pod, error) {
+	pod, err := podClient.Get(name, meta.GetOptions{})
+	if errors.IsNotFound(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve deleting pod: %v", err)
+	}
+	if pod.UID != uid {
+		return nil, nil
+	}
+	return pod, nil
+}
+
+func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
+	timeout := 300
+	pod, err := getPodByName(podClient, name, uid)
+	if pod != nil {
```

Comment:
Uh, `err != nil`?

Reply:
nope, we have `return nil` in some cases (e.g. wrong UID) when pod is gone

Reply:
yeah, but in the happy path we would get `pod, nil`, and would return instead of waiting for deletion.

Reply:
Reworked this in 93f52a7